cyberdrop_dl/crawlers/_tubecorporate.py

```
@@ -0,0 +1,85 @@
+        if await self.check_complete_from_referer(scrape_item):
+            return
+
+        api_url = self._get_api_url(scrape_item, video_id)
+        video = _get_default_video(await self.request_json(api_url))
+        video_info = await self._get_video_info(scrape_item, video_id)
+        scrape_item.possible_datetime = self.parse_iso_date(video_info["post_date"])
+
+        decoded_url = _decode_base64(video["video_url"])
+        link = self.parse_url(decoded_url, trim=False)
+        if not decoded_url.startswith("https"):
+            link = link.with_host(scrape_item.url.host)
+
+        filename, ext = self.get_filename_and_ext(link.parts[-2])
+        custom_filename = self.create_custom_filename(video_info["title"], ext, file_id=video_id)
+        return await self.handle_file(
+            scrape_item.url,
+            scrape_item,
+            filename,
+            ext,
+            custom_filename=custom_filename,
+            debrid_link=link,
+        )
+
+    async def _get_video_info(self, scrape_item: ScrapeItem, video_id: str) -> dict[str, str]:
+        json_url = self._get_json_url(scrape_item, video_id)
+        video_info: dict[str, str] = await self.request_json(json_url)
+        return video_info["video"]
+
+    def _get_json_url(self, scrape_item: ScrapeItem, video_id: str) -> AbsoluteHttpURL:
+        slug = f"{int(1e6 * (int(video_id) // 1e6))}/{1000 * (int(video_id) // 1000)}"
+        return scrape_item.url.with_path(f"api/json/video/86400/{slug}/{video_id}.json")
+
+    def _get_api_url(self, scrape_item: ScrapeItem, video_id: str) -> AbsoluteHttpURL:
+        query = {"video_id": video_id, "lifetime": "8640000"}
+        return scrape_item.url.with_path("api/videofile.php").with_query(query)
+
+
+def _get_default_video(formats: list[dict[str, str]]) -> dict[str, str]:
+    for fmt in formats:
+        if fmt.get("default") == "1":
+            return fmt
+    return formats[0]
```

For what i tested, every video has a single format and they have a key "is_default", not "default".

Did you find a video with more that one format? How do you know the default is always the best quality?

---

I haven't found a video with more than one format, just like you. You are right, the field is called is_default and it is a typing mistake. It was working because of the fallback returning always the first. I've added this check in case a video with more than one format exist, but I'm not sure it is needed.

---

I think is better to just throw an error if the video has more that 1 format.

We can add logic to handle it later if an user reports it, if they even exists.